src/tbb/private_server.cpp

```diff
@@ -351,49 +352,58 @@ inline bool private_server::try_insert_in_asleep_list( private_worker& t ) {
         return false;
     // Contribute to slack under lock so that if another takes that unit of slack,
     // it sees us sleeping on the list and wakes us up.
-    int k = ++my_slack;
-    if( k<=0 ) {
-        t.my_next = my_asleep_list_root.exchange(&t, std::memory_order_relaxed);
-        return true;
-    } else {
-        --my_slack;
-        return false;
+    auto expected = my_slack.load(std::memory_order_relaxed);
+    while (expected < 0) {
+        if (my_slack.compare_exchange_strong(expected, expected + 1)) {
+            t.my_next = my_asleep_list_root.load(std::memory_order_relaxed);
+            my_asleep_list_root.store(&t, std::memory_order_relaxed);
+            return true;
+        }
     }
+
+    return false;
 }
 
 void private_server::wake_some( int additional_slack ) {
-    __TBB_ASSERT( additional_slack>=0, NULL );
+    __TBB_ASSERT( additional_slack>=0, nullptr );
     private_worker* wakee[2];
     private_worker**w = wakee;
+
+    if (additional_slack) {
+        // Contribute our unused slack to my_slack.
+        my_slack += additional_slack;
+    }
+
+    int allotted_slack = 0;
+    while (allotted_slack < 2) {
+        // Chain reaction; Try to claim unit of slack
+        int old = my_slack.load(std::memory_order_relaxed);
+        do {
+            if (old <= 0) goto done;
+        } while (!my_slack.compare_exchange_strong(old, old - 1));
+        ++allotted_slack;
+    }
+done:
+
+    if (allotted_slack)
     {
         asleep_list_mutex_type::scoped_lock lock(my_asleep_list_mutex);
-        while( my_asleep_list_root.load(std::memory_order_relaxed) && w<wakee+2 ) {
-            if( additional_slack>0 ) {
-                // additional demand does not exceed surplus supply
-                if ( additional_slack+my_slack.load(std::memory_order_acquire)<=0 )
-                    break;
-                --additional_slack;
-            } else {
-                // Chain reaction; Try to claim unit of slack
-                int old = my_slack;
-                do {
-                    if( old<=0 ) goto done;
-                } while( !my_slack.compare_exchange_strong(old,old-1) );
-            }
+
+        while( my_asleep_list_root.load(std::memory_order_relaxed) && w<wakee+2 && allotted_slack) {
+            --allotted_slack;
             // Pop sleeping worker to combine with claimed unit of slack
             auto old = my_asleep_list_root.load(std::memory_order_relaxed);
             my_asleep_list_root.store(old->my_next, std::memory_order_relaxed);
```

Comment:
I believe we can be even better here by saving one more store to `my_asleep_list_root`, that is update it only once, despite the value of `allotted_slack`.

Reply:
I think you will still need a store to a local variable (maybe optimized into a register) to go through `old->my_next`. So, the difference is only between the store in the register + store to memory or two stores into the same memory. I would not expect notable performance difference because CPU is quite smart to optimize the stores into the same memory location.